### PR TITLE
feat: Add service request history for asset

### PR DESCRIPTION
**Pull request by Ashesh3**
Required for: https://github.com/coronasafe/care_fe/issues/5363
This pull request adds a new `AssetService` model to the project, which will allow us to store all records of service requests for each asset. This pull request also introduces an `AssetService` serializer in the `care/facility/api/serializers/asset.py` file. The `AssetService` serializer is used to retrieve the last service of an Asset. In the `AssetSerializer` serializer, `last_service` is added as a read-only field, and `last_serviced_on` and `note` are added as write-only fields. The `last_service` field retrieves the last service of the Asset using the `AssetService` serializer. 

This pull request also includes changes to the `create` and `update` methods of the `AssetSerializer` serializer to save the `last_service` data when an Asset is created or updated. The `AssetServiceFilter` filter is also added to the `care/facility/api/viewsets/asset.py` file to filter Assets by external id.

Migration has been created to move existing service request to a new AssetService object linked to that asset.
 
@coronasafe/code-reviewers

### CODE REVIEW

File: care/facility/migrations/0351_assetservice.py
Comment (sainak) [inline care/facility/migrations/0351_assetservice.py:15]: use bulk create

File: care/facility/api/serializers/asset.py
Comment (sainak) [inline care/facility/api/serializers/asset.py:153]: wrap this in a transaction

**Review by sainak — CHANGES_REQUESTED** [review]
run this branch through precommit

**Review by sainak — APPROVED** [review]
LGTM, squash the migrations after abdm is merged

**Comment by Ashesh3** [comment]
Lint won't run on the auto generated squashed migration file.

File: care/facility/api/viewsets/asset.py
Comment (vigneshhari) [inline care/facility/api/viewsets/asset.py:321]: Are we okay showing service history without any authZ

**Comment by vigneshhari** [comment]
Squash all migrations to a single one.

**Comment by vigneshhari** [comment]
@rithviknishad add your review.

File: care/facility/api/viewsets/asset.py
Comment (Ashesh3) [reply care/facility/api/viewsets/asset.py:321]: @vigneshhari If I remember correctly, the asset details are publicly available and are shown on the state dashboard, in that case it should be fine to make their service history public too, right? 

cc: @gigincg

Comment (Ashesh3) [reply care/facility/api/viewsets/asset.py:321]: @vigneshhari Aparna confirmed to have it publicly visible.

File: care/facility/models/asset.py
Comment (rithviknishad) [inline care/facility/models/asset.py:190]: Do we need to allow asset service records without a serviced_on datetime for the record?

**Review by rithviknishad — COMMENTED** [review]
rebase and squash the migrations btw

File: care/facility/models/asset.py
Comment (Ashesh3) [reply care/facility/models/asset.py:190]: Currently the system allows users to enter a blank date with a note or vice versa, so I carried over that behavior. It doesn't make sense to have a service record without a date, but if it's being used already then, we have to keep the behavior, otherwise we can make the date mandatory.

Comment (rithviknishad) [reply care/facility/models/asset.py:190]: so the timeline table would be sorted by? modified/created date?

Comment (rithviknishad) [reply care/facility/models/asset.py:190]: @gigincg is it fine to allow serviced_on to be blank?

we could just allow null for migration purposes but have `blank=False` to make it a required field in forms.

Comment (Ashesh3) [reply care/facility/models/asset.py:190]: okay, let's do that!

Comment (Ashesh3) [inline care/facility/models/asset.py:190]:
```suggestion
    serviced_on = models.DateField(default=None, null=True, blank=False)
```

File: care/facility/api/serializers/asset.py
Comment (rithviknishad) [inline care/facility/api/serializers/asset.py:160]:
```suggestion
                asset_instance.save(update_fields=["last_service"])
```

**Review by rithviknishad — CHANGES_REQUESTED** [review]
Tests are failing; you'll need to makemigrations again 😅

**Review by rithviknishad — APPROVED** [review]

**Review by sainak — APPROVED** [review]

File: care/facility/api/serializers/asset.py
Comment (sainak) [inline care/facility/api/serializers/asset.py:86]:
```suggestion
```
this is not required anymore

File: care/facility/api/serializers/asset.py
Comment (sainak) [inline care/facility/api/serializers/asset.py:110]:
```suggestion
                serviced_on=serviced_on,
                note=note,
```

**Review by sainak — CHANGES_REQUESTED** [review]

**Review by sainak — APPROVED** [review]

File: care/facility/models/asset.py
Comment (rithviknishad) [inline care/facility/models/asset.py:193]: Why is this many to many?

**Review by rithviknishad — APPROVED** [review]

File: care/facility/models/asset.py
Comment (sainak) [inline care/facility/models/asset.py:193]:
```suggestion
```

File: care/facility/models/asset.py
Comment (sainak) [inline care/facility/models/asset.py:202]:
```suggestion
        AssetService, on_delete=models.CASCADE, null=False, blank=False, related_name="edits"
```

File: care/facility/api/serializers/asset.py
Comment (sainak) [inline care/facility/api/serializers/asset.py:90]:
```suggestion
    edits = AssetServiceEditSerializer(many=True)
```

File: care/facility/api/viewsets/asset.py
Comment (sainak) [inline care/facility/api/viewsets/asset.py:361]:
```suggestion
        )
        .prefetch_related("edits")
        .order_by("-created_date")
```

**Review by sainak — CHANGES_REQUESTED** [review]

File: care/facility/models/asset.py
Comment (Ashesh3) [reply care/facility/models/asset.py]: image

Resolved.

**Comment by sainak** [comment]
@Ashesh3 migrations needs to updated

**Review by sainak — APPROVED** [review]

**Review by rithviknishad — APPROVED** [review]

**Review by sainak — APPROVED** [review]

**Review by rithviknishad — APPROVED** [review]

**Review by vigneshhari — APPROVED** [review]

**Pull request merged**
